akka-typed/src/main/scala/akka/typed/javadsl/BehaviorChain.scala

```diff
@@ -0,0 +1,243 @@
+/**
+ * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.typed.javadsl
+
+import scala.annotation.tailrec
+import akka.japi.function.{ Function, Function2, Predicate }
+import akka.typed
+import akka.typed.{ Behavior, ExtensibleBehavior, Signal }
+import BehaviorChain._
+import akka.annotation.InternalApi
+
+/**
+ * Used for creating a [[Behavior]] by 'chaining' message and signal handlers.
+ *
+ * When handling a message or signal, this [[Behavior]] will consider all handlers in the order they were added,
+ * looking for the first handler for which both the type and the (optional) predicate match.
+ *
+ * @tparam T the common superclass of all supported messages.
+ */
+class BehaviorChain[T] private (
+  private val messageHandlers: List[Case[T, T]],
+  private val signalHandlers:  List[Case[T, Signal]]
+) extends ExtensibleBehavior[T] {
+
+  override def receiveMessage(ctx: typed.ActorContext[T], msg: T): Behavior[T] = receive[T](ctx, msg, messageHandlers)
+
+  override def receiveSignal(ctx: typed.ActorContext[T], msg: Signal): Behavior[T] = receive[Signal](ctx, msg, signalHandlers)
+
+  @tailrec
+  private def receive[M](ctx: ActorContext[T], msg: M, handlers: List[Case[T, M]]): Behavior[T] =
+    handlers match {
+      case Case(cls, predicate, handler) :: tail ⇒
+        if (cls.isAssignableFrom(msg.getClass) && (predicate.isEmpty || predicate.get.apply(msg))) handler(ctx, msg)
+        else receive[M](ctx, msg, tail)
+      case _ ⇒
+        // emulate scala match error
+        throw new MatchError(msg)
+    }
+
+  /**
+   * Add a new case to the message handling.
+   *
+   * @param type type of message to match
+   * @param handler action to apply if the type matches
+   * @tparam M type of message to match
+   * @return a new behavior with the specified handling appended
+   */
+  def message[M <: T](`type`: Class[M], handler: Function2[ActorContext[T], M, Behavior[T]]): BehaviorChain[T] =
+    new BehaviorChain[T](messageHandlers :+ Case[T, T](`type`, None, (i1: ActorContext[T], msg: T) ⇒ handler.apply(i1, msg.asInstanceOf[M])), signalHandlers)
```

Comment:
use `Vector` if appending, `List` is for prepending, otherwise we could use real builder and reverse